lib/puppet/etc.rb

```
@@ -0,0 +1,118 @@
+    # Etc::getpwnam searches /etc/passwd file for an entry corresponding to
+    # username.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwnam(username)
+      if user_entry = Etc.getpwnam(username)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    # Etc::getgrnam searches /etc/group file for an entry corresponding to groupname.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrnam(groupname)
+      if group_entry = Etc.getgrnam(groupname)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getgrid searches /etc/group file for an entry corresponding to id.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrgid(id)
+      if group_entry = Etc.getgrgid(id)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getpwuid searches /etc/passwd file for an entry corresponding to id.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwuid(id)
+      if user_entry = Etc.getpwuid(id)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    private
+    # utility method for converting the values of a struct returned by the Etc
+    # module to UTF-8.
+    #
+    # Warning! This is a destructive method - the struct passed is modified!
+    #
+    # Structs returned by the ruby Etc module contain members with
+    # fields of type String, Integer, or Array of Strings, so we handle these
+    # types. Otherwise ignore fields.
+    def convert_field_values_to_utf8!(struct)
+      struct.each_with_index do |value, index|
+        if value.is_a?(String)
+          struct[index] = Puppet::Util::CharacterEncoding.convert_to_utf_8(value)
```

Between partially managing a user's properties and failing to apply the resource entirely, the latter option is 1) visible and 2) less dangerous. Failing early on bad input means we won't leave a system in a partially managed, but mangled, state.